### PR TITLE
fix: grid selection column indeterminate state (#2954) (CP: 14.8)

### PR DESCRIPTION
**Pull request by sissbruecker**
Cherry-pick of: #2954 

(cherry picked from commit 08fcccf06b733d4760d66bf83f840cd20edd0af1)

### CODE REVIEW

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/vaadin-grid-flow-selection-column.html [hunk 1]
```diff
@@ -3,12 +3,19 @@
 <dom-module id="vaadin-grid-flow-selection-column">
   <template>
     <template class="header" id="defaultHeaderTemplate">
+      <style>
+          /* Fix a grid web-component style that sets the font-size to small for all header contents */
+          #selectAllCheckbox {
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/vaadin-grid-flow-selection-column.html:8]: This is a new addition as it only applies to v14. It addresses a minor comment from https://github.com/vaadin/flow-components/issues/1060:

> Another minor flaw is that the checkbox in header appears to be smaller than checkboxes in column.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/vaadin-grid-flow-selection-column.html [hunk 2]
```diff
@@ -3,12 +3,19 @@
 <dom-module id="vaadin-grid-flow-selection-column">
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/vaadin-grid-flow-selection-column.html:3]: v14 has duplicate implementations of the grid selection column for Polymer 3, and for Polymer 2 (Bower mode). Applied the same changes to the Polymer 2 version of the component here.

Apart from that the component, and the changes here, have diverged quite a bit from v23, but the changes are mostly simple.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModelTest.java [hunk 3]
```diff
@@ -95,18 +97,194 @@ public void singleItemSignature_deselectFormClient_dontSendToClientSide() {
...
+    }
+
+    @Test
+    public void updateSelection_updatesCheckboxStates() {
+        grid.setSelectionMode(SelectionMode.MULTI);
+        grid.setItems("foo", "bar");
+        Element columnElement = getGridSelectionColumn(grid).getElement();
+
+        // Select all
+        grid.asMultiSelect().updateSelection(new HashSet<>(Arrays.asList("foo", "bar")), new HashSet<>());
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModelTest.java:263]: `Set.of` was not availabe in Java 8, switched to constructing hash sets.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/MockUI.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+import com.vaadin.flow.component.UI;
+import com.vaadin.flow.server.VaadinRequest;
+import com.vaadin.flow.server.VaadinService;
+import com.vaadin.flow.server.VaadinSession;
+
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class MockUI extends UI {
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/MockUI.java:11]: The v23 version uses a public class from a test in the Flow framework itself (`com.vaadin.flow.data.provider.DataCommunicatorTest`), which does not exist in v14. Copied the class over, which seems to work fine.

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridViewIT.java [hunk 5]
```diff
@@ -285,17 +285,25 @@ public void gridAsMultiSelect() {
...
         // test the select all button
-        grid.findElement(By.id("selectAllCheckbox")).click();
+        TestBenchElement selectAllCheckbox = grid
+                .findElement(By.id("selectAllCheckbox"));
+        selectAllCheckbox.click();
         // deselect 1
         getCellContent(grid.getCell(0, 0)).click();
-        Assert.assertEquals("Select all should have been deselected", "false",
-                grid.findElement(By.id("selectAllCheckbox"))
-                        .getAttribute("aria-checked"));
```
Comment (sissbruecker) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/grid/it/GridViewIT.java:293]: I guess this test was removed / refactoring in newer versions, updated the assertions here to include indeterminate state.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java [hunk 6]
```diff
@@ -113,9 +113,18 @@ public void selectFromClient(T item) {
             fireSelectionEvent(new MultiSelectionEvent<>(getGrid(),
                     getGrid().asMultiSelect(), oldSelection, true));
 
-            long size = getDataProviderSize();
+            if (!isSelectAllCheckboxVisible()) {
+                // Skip changing the state of Select All checkbox if it was
```
Comment (vursen) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:117]: I only noticed now that this condition is absent in all the other methods concerning the selection. Why? If it is an optimization, I would expect it to be in all the methods. This question also comes to the original PR for V23.

Comment (sissbruecker) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java:117]: I though about this as well, but it was like that before and the intention of the PR is to add the indeterminate state, not to optimize performance, so I left as is.

**Review by vursen — APPROVED** [review]

**Pull request merged**